Add support for canvas and DOM rendering to `Legend`

mattpap opened the pull request:
The goal of this PR is allow both canvas and DOM/CSS rendering for renderers for which it makes sense, i.e. renderers that are UI components and renderers that include text. Secondary goal is to replace custom layouts with CSS layout where every possible without a major redesign. I expect only a selected subset of renderers will be covered in this PR. I started with `Legend` because it touches every aspect that I need to cover by the new design.

The effects of this PR will be:

* better integration with DOM/CSS (e.g. responsiveness to font loading/changes)
* better theming and customizability beyond what visuals offer (no support for save/copy tool in this case)
* support for text selection without the need for any custom implementations and new tools
* ease of development (renderers are inspectable in devtools)
* anything related to accessibility

[Screencast from 16.08.2024 11:49:27.webm](https://github.com/user-attachments/assets/6503c161-8640-4477-a352-4b1673e99d1e)

- [x] addresses #10299 
- [x] fixes #13566 (regression test added)
- [x] addresses #10205
- [x] fixes #6662

mattpap committed:
Add support for dual canvas and DOM/CSS rendering

bryevdv commented:
@mattpap should the linked issues be triaged as `feature`? As an aside, I'd make a pitch for colorbar to be the next target for this kind of treatment.

mattpap commented:
>  should the linked issues be triaged as feature?

Yes. I'm actually triaging issues that may be affected by this work.

> As an aside, I'd make a pitch for colorbar to be the next target for this kind of treatment.

Axes and title will have to be next. Then it will be possible to work on color bars.

mattpap committed:
Allow rendering_target() to return shadow root

mattpap committed:
Add CartesianFrame's element to events layer

mattpap committed:
Allow recursive View.on_transitive_change()

mattpap committed:
Remove LegendItem.legend backref and use modern API

mattpap committed:
Update Legend's integration tests

mattpap committed:
Move positioning logic to CanvasPanel

mattpap committed:
Add CanvasPanels for side panels for positioning

mattpap committed:
Allow Legend to target CanvasPanel side panels

mattpap committed:
Update Legend's bbox in time for the layout

mattpap committed:
Apply high DPI transform when painting glyphs

mattpap committed:
Don't let computed stylsheets contents accumulate

mattpap committed:
Allow to rotate title despite limitations of CSS

mattpap committed:
Better organization of different kinds of stylesheets

mattpap committed:
Use max-content for Legend's grid track sizes

mattpap committed:
Use relative bbox for [x, y] positioning

mattpap committed:
Restore support for label_{width,height}

mattpap committed:
Don't use raw CSS classes in computed stylesheets

mattpap committed:
Allow renderers to paint to custom canvas layers

mattpap committed:
Export canvas via repainting and not compositing

mattpap committed:
Restore support for painting Legend to canvas

mattpap commented:
Legend with DOM (left) and canvas rendering (right) side-by-side:
![Screenshot from 2024-09-17 13-45-09](https://github.com/user-attachments/assets/6a868f5f-4e9c-47e4-869b-c5f6e04f3927)

mattpap commented:
I was hoping to finish this PR as a big feature for bokeh 3.6, but that's unrealistic given how much additional work needed to be done to support DOM rendering and how much more is needed to tweak and fine tune the implementation. What needs to be done is:

- [x] fine tune CSS to cover all the corner cases (e.g. angled text)
- [x] address CSS cross-browser compatibility issues (e.g. vertical text in FF)
- [x] finalize support for positioning in side panels
- [x] don't render and paint at the same time (new output backend?)
- [x] address performance issues due to over-rendering (due to uncertainty when to update the legend)

mattpap committed:
Add a stylesheet to CanvasPanel

mattpap committed:
Temporarily exclude CanvasPanel from serializable state

mattpap committed:
Either paint or render but not both

mattpap committed:
Resolve cross-browser issues related to vertical text

mattpap committed:
Add release notes

mattpap committed:
CSSify center and side panel layouts

mattpap committed:
Invalidate layout when Legend resizes

mattpap committed:
Address over-rendering performance issues

mattpap committed:
Update SVG export unit tests

mattpap committed:
Make sure users can override any stylesheets

mattpap committed:
Use panel relative positioning in HTMLLabel

mattpap committed:
Don't attempt to paint glyphs prematurely

mattpap committed:
Make sure render() is followed by layout

mattpap commented:
There is still one regression to address, but otherwise this is ready for review. Note that great many changes to the internals are more or less temporary, and will get refined in future PRs touching the subject. These will include changes to axes, labels, title and text glyphs.

mattpap committed:
Improve configuration of canvas panels

mattpap committed:
Help CSS layout with rotated labels in panels

bryevdv commented on bokehjs/test/baselines/linux/Legend_annotation__in_vertical_orientation__should_support_grid_layout__with_nrows=auto_and_ncols=3.blf:
```diff
@@ -21,5 +21,5 @@ Figure bbox=[0, 0, 300, 300]
-  Legend bbox=[49, 15, 236, 83]
-  Legend bbox=[49, 185, 236, 83]
+  Legend bbox=[56, 45, 239, 85]
+  Legend bbox=[56, 213, 239, 85]
```
Commenting here since I can't comment on the image. The `fill=even` and `fill=odd` legends on these tests look identical. Are they supposed to differ at all in some way?

mattpap replied:
I forgot to compute canvas relative bbox out of CSS positioning, which is viewport relative.

mattpap replied:
After fixing this, any additional differences a result of:

1. Bokeh's layout not taking borders into consideration.
2. Slight differences in text rendering and alignment between CSS and our text rendering engine.

bryevdv commented on src/bokeh/models/annotations/legends.py:
```diff
@@ -489,9 +489,14 @@ def __init__(self, *args, **kwargs) -> None:
-    Amount of padding around the contents of the legend. Only applicable when
-    border is visible, otherwise collapses to 0.
+    padding = Padding(default=10, help="""
+    Amount of padding around the contents of the legend.
+
+    Only applicable when border is visible, otherwise collapses to 0.
```
```suggestion
    Amount of padding between the contents of the legend and the legend border.

    Only applicable when border is visible.
```

bryevdv commented on bokehjs/src/lib/models/widgets/tables/data_table.ts:
```diff
@@ -183,16 +183,10 @@ export class DataTableView extends WidgetView {
-        this.render()
-        this.r_after_render()
-      })
+      this.connect(column.change, () => this.rerender())
```
Is `renderer` a good name here? Its a noun, not a verb, so I'd expect it e.g. to return some object or thing, not "do" something (which is what I would expect an event callback to do)

mattpap replied:
I'm pretty sure "re-render" is a verb, but I suppose it depends how you use it in a sentence. To me it's clear this is an action to take, the same as "redo", etc.

bryevdv replied:
Sorry I literally just misread it since "renderER" is already so common in the codebase. `re_render` would help but is a bit ugly I guess. Maybe it's fine as-is.

bryevdv commented:
Still reviewing but a few initial comments. All the image diff changes looked very minimal / fine except one thing that jumped out at me. The item label baseline looks shifted up from the center of the item glyph. I'm not sure it was ever actually centered but the new versions are now _just_ high enough that they look slightly wrong to me. See e.g. below the bottom horizontal line in `#1` is just below the orange line in the glyph in the "before" and just above the orange line in "after".

<img width="1400" alt="Screenshot 2024-10-24 at 11 21 31" src="https://github.com/user-attachments/assets/d5098963-ea0b-4501-b782-0cf5f662eae1">


Is this intentional or unavoidable? Alternatively, if this is "correct" by some measure is there a new different default we should apply to get things more horizontally centered?

bryevdv commented on bokehjs/src/lib/core/dom_view.ts:
```diff
@@ -53,6 +53,11 @@ export abstract class DOMView extends View {
 
+  rerender(): void {
+    this.render()
+    this.r_after_render()
```
What does the "r" designate?

mattpap replied:
`r` stands for `recursive`, though this isn't currently a consistent naming scheme. E.g. `update_layout()` is recursive, but not indicated as such. Typically `r_` methods will appear where there's already a non-recursive method of the same name.

bryevdv commented on bokehjs/src/lib/core/util/bbox.ts:
```diff
@@ -468,6 +473,11 @@ export class BBox implements Rect, Equatable {
+  relative_to(to: BBox): BBox {
+    const {x, y, width, height} = this
+    return new BBox({x: x - to.x, y: y - to.y, width, height})
+  }
```
these seem amenable to unit test

bryevdv commented on bokehjs/src/lib/core/util/canvas.ts:
```diff
@@ -57,7 +58,13 @@ export class CanvasLayer {
+        const {x, y, width, height} = bbox
+        this.rect(x, y, width, height)
+      },
+    })
```
seems a little weird to patch like this, is this expected to go away at some point?

mattpap replied:
This is quite standard practice, as long as it doesn't conflict with existing APIs. What would be somewhat weird, it to override `rect()` with custom logic and API overrides. The alternative to this would be to create a new class and proxy existing methods and attributes manually, which would be a lot of code. It has to be done manually btw., because real proxies in JS are awfully slow.

bryevdv commented on bokehjs/src/lib/models/annotations/title.ts:
```diff
@@ -22,7 +20,7 @@ export class TitleView extends TextAnnotationView {
     const {bbox} = this.layout
 
-    switch (this.panel.side) {
+    switch (this.panel!.side) {
```
what is needed to get rid of the `!` ?

mattpap replied:
This will eventually go away. Previously these models had `declare panel: SidePanel` to remove the `undefined` case inherited from a base class. Due to JS/TS technicalities around overriding attributes with getters/setters, I had to remove those declarations, which results in having to use `!` operator.

mattpap committed:
Compute Legend's bbox relative to canvas

mattpap committed:
Update visual baselines

mattpap commented:
> I'm not sure it was ever actually centered but the new versions are now just high enough that they look slightly wrong to me. See e.g. below the bottom horizontal line in #1 is just below the orange line in the glyph in the "before" and just above the orange line in "after".

They are centered by CSS standards, however they look kind of offset in these images. They look closer to the original when run manually in a non-headless setup. Some minor differences in rendering are expected, because we do font/text measurements slightly differently from what CSS does. Maybe those images are captured too early, before CSS is done with layout. UI elements accommodate for that, but `Legend` isn't a UI element yet, so I guess I will have to deal with that sooner than later.

mattpap committed:
Correct title location in painted output

mattpap committed:
Add image export tests for Legend

mattpap committed:
Update visual baselines

mattpap committed:
Fix HTMLLabel after switching to CSS layout

mattpap commented:
This is now fully ready. I finished `HTMLLabel`'s rendering implementation update, fixed a few issues with `Legend`'s implementation and added tests for image export related to `Legend`.

bryevdv commented:
@mattpap one of the images at least seems to have an issue?

<img width="1377" alt="Screenshot 2024-10-29 at 12 00 46" src="https://github.com/user-attachments/assets/ec2cf9cc-36f6-4be5-b426-143dc5da9cd0">

mattpap committed:
Fix vertical positioning of items' labels

mattpap commented:
> one of the images at least seems to have an issue?

It's fixed now.

bryevdv reviewed (approved):
I didn't have a chance to go over the the JS super carefully, but I did not see any obvious issues. I did go over the images quite carefully. @mattpap if you'd like more code review I'd encourage you to request from @philippjfr or @ianthomas23 directly. Otherwise, probably just better to get this merged earlier to see how it shakes out.

mattpap commented:
Test failure is unrelated and reported in issue #14128. I'm going ahead and merge this. There is still a lot of temporary stuff in this PR or stuff that require generalization, that when refined in hopefully in smaller PRs, will be easier for reviewers to digest.

gabalafou commented:
I see that this has been merged and I have only given it a cursory look so far.

I was asked to check this work in Safari specifically. So I built the js and the docs with this PR applied and the first plot that I opened ([transform markers](https://docs.bokeh.org/en/latest/docs/examples/basic/data/transform_markers.html)) I noticed a pretty big difference between Chrome versus Safari.

Chrome:

<img width="618" alt="image" src="https://github.com/user-attachments/assets/1d99a350-094e-4e2f-bb26-5d26e199032f">

Safari:

<img width="614" alt="image" src="https://github.com/user-attachments/assets/05c1a5ee-ce75-49ad-b58f-09ad07a0fdf4">

mattpap commented:
@gabalafou, I can reproduce this in epiphany (webkit on Linux). Can you start an issue for this? Although I can reproduce the issue, that browser is insufferable and I can't inspect or debug anything in it at this point at least.

gabalafou commented:
I've started an issue here: https://github.com/bokeh/bokeh/issues/14134

I added the `plat: safari` and `type: bug` labels but I'm not entirely sure how to title and label this issue.

mattpap commented:
Thanks.

> I added the `plat: safari` and `type: bug` labels but I'm not entirely sure how to title and label this issue.

That's a good start.
